mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,819 @@
+                }
+
+                return Self::change_state(stopped);
+            },
+        };
+
+        let last_fetched_block = cmp::max(highest_send_tx_height, highest_recieved_tx_height);
+
+        log::info!(
+            "Tx history fetching finished for {}. Last fetched block {}",
+            ctx.coin.platform_ticker(),
+            last_fetched_block
+        );
+
+        ctx.coin.set_history_sync_state(HistorySyncState::Finished);
+        Self::change_state(WaitForHistoryUpdateTrigger::new(
+            self.address.clone(),
+            last_fetched_block,
+        ))
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> State for TendermintInit<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        const INITIAL_SEARCH_HEIGHT: u64 = 0;
+
+        ctx.coin.set_history_sync_state(HistorySyncState::NotStarted);
+
+        if let Err(e) = ctx.storage.init(&ctx.coin.history_wallet_id()).await {
+            return Self::change_state(Stopped::storage_error(e));
+        }
+
+        let search_from = match ctx
+            .storage
+            .get_highest_block_height(&ctx.coin.history_wallet_id())
+            .await
+        {
+            Ok(Some(height)) if height > 0 => height as u64 - 1,
+            _ => INITIAL_SEARCH_HEIGHT,
+        };
+
+        Self::change_state(FetchingTransactionsData::new(
+            ctx.coin.my_address().expect("my_address can't fail"),
+            search_from,
+        ))
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> LastState for Stopped<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> Self::Result {
+        log::info!(
+            "Stopping tx history fetching for {}. Reason: {:?}",
+            ctx.coin.ticker(),
+            self.stop_reason
+        );
+
+        let new_state_json = json!({
+            "message": format!("{:?}", self.stop_reason),
+        });
+
+        ctx.coin.set_history_sync_state(HistorySyncState::Error(new_state_json));
+    }
+}
+
+#[async_trait]
+impl TendermintTxHistoryOps for TendermintCoin {
+    async fn get_rpc_client(&self) -> MmResult<HttpClient, TendermintCoinRpcError> { self.rpc_client().await }
+
+    fn decimals(&self) -> u8 { self.decimals }
+
+    fn platform_denom(&self) -> String { self.denom.to_string() }
+
+    fn set_history_sync_state(&self, new_state: HistorySyncState) {
+        *self.history_sync_state.lock().unwrap() = new_state;
+    }
+
+    async fn all_balances(&self) -> MmResult<AllBalancesResult, TendermintCoinRpcError> { self.all_balances().await }
```

Comment:
I personally think that `all_balances` is related to the history operating, so I'd leave it as it is :)
But about `get_rpc_client` - makes sense, it could be placed in `tendermint.rs`, for example

Reply:
> I personally think that all_balances is related to the history operating

I agree, this is probably the right place for `all_balances`. I guess `all_balances` in `tendermint_coin.rs`https://github.com/KomodoPlatform/atomicDEX-API/blob/e5308d97dda876e1d8adfc528667918dc56f957c/mm2src/coins/tendermint/tendermint_coin.rs#L631 should be removed and implemented here directly instead to avoid having the same function twice.

Reply:
This is used for coin activation as well.

Reply:
Then `all_balances` also could be move to a separate trait :)